### DISCUSSION TITLE
Update bimmer_connected to 0.14.3  and upgrade to Python 3.8.4 - module with error message

### DISCUSSION THREAD

**Discussion by Isy99** [bimmerconnected/bimmer_connected#582] [General]
Hello team,
todays my installation added big list of failures to the log of my Homeautomation system (i.E. FHEM per mqtt). Since April this year everything was fine.

I decided to install bimmer_connected to 0.14.3  and upgraded to Python 3.8.4:
```
pi@fhem:~ $ python3 -V
Python 3.8.4
```

When I start the script from CLI I get the following response:

```
Client MyBMW sending CONNECT
Client MyBMW received CONNACK (0)
Traceback (most recent call last):
  File "/usr/local/bin/bimmerconnected", line 6, in <module>
    from bimmer_connected.cli import main
  File "/usr/local/lib/python3.7/dist-packages/bimmer_connected/__init__.py", line 12, in <module>
    from importlib.metadata import version
ModuleNotFoundError: No module named 'importlib.metadata'
Client MyBMW sending DISCONNECT
```

It looks like somehow Python 3.7 still is used.
Can someone from you help or give advice, what to do?

Thanks in advance,
Helmut

Just installed Python 3.11 and tried to install bimmer_connected.
Result is a FM 
```
Failed to build Pillow
ERROR: Could not build wheels for Pillow, which is required to install pyproject.toml-based projects
```

No idea what to do!

**Comment by rikroe** [comment]
Hi,

yes, Python 3.7 is not supported anymore due to changes in Pythons standard lib (and HomeAssistant, our main user being quite strict with deprecations).

Are you using a RPi as server? If yes, which one? And could you provide the output of `uname -m` ?

**Comment by Isy99** [comment]
Hello Richard!
Thank you very much for your help! I've somehow background in IT, but with this problem I'm totally lost.

Yes, Python 3.11 etc is installed on Pi 4B, OS Buster.
uname -m returns
`armv71`

Just checked PIP version
pip -V returns
`pip 23.3.1 from /home/pi/.local/lib/python3.8/site-packages/pip (python 3.8)`
What seems to be wrong, as it points to Python 3.8 (a former install)
I uninstalled pip via
`python3 -m pip uninstall pip`

Tried to re-install latest pip3
`python -m ensurepip --upgrade`

pip -V returns
`pip 22.3 from /home/pi/.local/lib/python3.11/site-packages/pip (python 3.11)`
Looks much better now!

Again install bimmer_connected:
`pip3 install --upgrade bimmer_connected`
No FM!
After install hint to upgrade pip3 to the latest version:
`python3.11 -m pip install --upgrade pip`
Worked as well!

Test with
`bimmerconnected status --json mail@domain.de bmw_passw rest_of_world `
returns FM
```
Traceback (most recent call last):
  File "/usr/local/bin/bimmerconnected", line 6, in <module>
    from bimmer_connected.cli import main
  File "/usr/lib/python3.11/site-packages/bimmer_connected/cli.py", line 14, in <module>
    from bimmer_connected.account import MyBMWAccount
  File "/usr/lib/python3.11/site-packages/bimmer_connected/account.py", line 11, in <module>
    from bimmer_connected.api.authentication import MyBMWAuthentication
  File "/usr/lib/python3.11/site-packages/bimmer_connected/api/authentication.py", line 18, in <module>
    from bimmer_connected.api.utils import (
  File "/usr/lib/python3.11/site-packages/bimmer_connected/api/utils.py", line 21, in <module>
    from PIL import Image
  File "/usr/lib/python3.11/site-packages/PIL/Image.py", line 82, in <module>
    from . import _imaging as core
ImportError: libtiff.so.6: cannot open shared object file: No such file or directory

```

I hope this info was helpful so far!

I'll try to support as much as I ca to get your nice tool working back again. It's in productive environment and I control photovoltaics based charging.

Kind regards,
Helmut

**Comment by rikroe** [comment]
Hi Helmut,

sometimes Python packages are installed in locations they should not be, i.e. user vs system store or a virtual env. Combined with the CLI registration, this sometimes causes confusion. But seems you figured that out :)

Regarding the Pillow issue, please check https://github.com/bimmerconnected/bimmer_connected/issues/579#issuecomment-1821865578 and isntall the required dependencies. 
We will also release a version without pillow soon, didn't think we get so many issues with it.

**Comment by Isy99** [comment]
Hi Richard,
right, my Raspbian is 32bit. 
The steps mentioned in https://github.com/bimmerconnected/bimmer_connected/issues/579#issuecomment-1821865578 did not really help.

```
sudo apt-get install libtiff5-dev libjpeg8-dev libopenjp2-7-dev zlib1g-dev \
    libfreetype6-dev liblcms2-dev libwebp-dev tcl8.6-dev tk8.6-dev python3-tk \
    libharfbuzz-dev libfribidi-dev libxcb1-dev
```

results in other  open dependencies.
```
Die folgenden Pakete haben unerfüllte Abhängigkeiten:
 libtiff-dev : Hängt ab von: libjpeg-dev
```
Added  libjpeg-dev results in:
```
 libjpeg-dev : Hängt ab von: libjpeg62-turbo-dev (>= 1:1.5.2-2+deb10u1) soll aber nicht installiert werden
               Kollidiert mit: libjpeg8-dev aber 8d1-2 soll installiert werden

```

Any idea?

Best regards, Helmut

**Comment by rikroe** [comment]
Puh, not really... (other than replacing the 32bit OS with 64bit)
As soon as https://github.com/bimmerconnected/bimmer_connected/pull/583 is merged, I'll create a new release that does not require Pillow anymore.

**Comment by Isy99** [comment]
Moin Richard,
sounds good to me. I'll be waiting for that version.

During Xmas I need to setup my HA system on a new PI with the latest OS, as I'm still on Buster. Much work..........I'm using the PIO ports and always requires special and different settings in the OS.

**Comment by rikroe** [comment]
Please try installing 0.14.4 as you would do normally - it should work now.

**Comment by Isy99** [comment]
**Yeah - that works fine again. Thank so very much.**

I still have problems with the path, as I start a shell file ("getBimmer.sh") from FHEM, I need to ask someone from the FHEM/Linux specialists.
"bimmerconnected" is now installed to /home/pi/.local/bin and the user "them" still starts the old version, although I changed the path in the getBimmer.sh script.

**Comment by rikroe** [comment]
Then you have installed bimmer-connected multiple times (maybe once using the "pi" user and once using the "fhem" user, and maybe even once as admin using sudo).

For each of the users (including once with sudo), do a pip uninstall bimmer-connected until nothing is installed anymore. Then install for fhem again.

Also, maybe fhem is using a virtual environment that you need to activate before updating bimmer connected?

**Comment by Isy99** [comment]
Hello Richard, thanks again for junping in.
I tried your hints, does not help.

Manually, user pi, I enter the following command, that works:
PATH=$PATH:/home/pi/.local/bin bimmerconnected status --json xxx@mai.de password rest_of_world | mosquitto_pub -h 192.168.178.86 -p 1883 -u is -P password -d -t /bmw -l -i MyBMW --quiet

Starting the same from FHEM returns an error:
```
Traceback (most recent call last):
  File "/home/pi/.local/bin/bimmerconnected", line 5, in <module>
    from bimmer_connected.cli import main
  File "/usr/lib/python3.11/site-packages/bimmer_connected/cli.py", line 14, in <module>
    from bimmer_connected.account import MyBMWAccount
  File "/usr/lib/python3.11/site-packages/bimmer_connected/account.py", line 11, in <module>
    from bimmer_connected.api.authentication import MyBMWAuthentication
  File "/usr/lib/python3.11/site-packages/bimmer_connected/api/authentication.py", line 18, in <module>
    from bimmer_connected.api.utils import (
  File "/usr/lib/python3.11/site-packages/bimmer_connected/api/utils.py", line 21, in <module>
    from PIL import Image
  File "/usr/lib/python3.11/site-packages/PIL/Image.py", line 82, in <module>
    from . import _imaging as core
ImportError: libtiff.so.6: cannot open shared object file: No such file or directory

```
Still pillow somehow involved?
I tried to uninstall pillow with "pip uninstall pillow", what ends in an error message.
I think, I installed pillow or pip with sudo before.......

```
pi@fhem:~ $ pip uninstall pillow
Found existing installation: Pillow 10.1.0
Uninstalling Pillow-10.1.0:
  Would remove:
    /usr/lib/python3.11/site-packages/PIL/*
    /usr/lib/python3.11/site-packages/Pillow-10.1.0.dist-info/*
Proceed (Y/n)? Y
ERROR: Exception:
Traceback (most recent call last):
  File "/usr/lib/python3.11/shutil.py", line 824, in move
    os.rename(src, real_dst)
PermissionError: [Errno 13] Keine Berechtigung: '/usr/lib/python3.11/site-packages/PIL/' -> '/tmp/pip-uninstall-_f12qfwg'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
    status = run_func(*args)
             ^^^^^^^^^^^^^^^
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/commands/uninstall.py", line 105, in run
    uninstall_pathset = req.uninstall(
                        ^^^^^^^^^^^^^^
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/req/req_install.py", line 725, in uninstall
    uninstalled_pathset.remove(auto_confirm, verbose)
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/req/req_uninstall.py", line 381, in remove
    moved.stash(path)
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/req/req_uninstall.py", line 272, in stash
    renames(path, new_path)
  File "/home/pi/.local/lib/python3.11/site-packages/pip/_internal/utils/misc.py", line 362, in renames
    shutil.move(old, new)
  File "/usr/lib/python3.11/shutil.py", line 842, in move
    rmtree(src)
  File "/usr/lib/python3.11/shutil.py", line 731, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/usr/lib/python3.11/shutil.py", line 682, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/usr/lib/python3.11/shutil.py", line 680, in _rmtree_safe_fd
    os.unlink(entry.name, dir_fd=topfd)
PermissionError: [Errno 13] Keine Berechtigung: 'ImageDraw2.py'
```

P.S. Do you have a donation link?

**Comment by rikroe** [comment]
While your cli tries to use the one in your user directory (`/home/pi/.local/bin/bimmerconnected`) it is using bimmer connected from the system installation (`/usr/lib/python3.11/site-packages/bimmer_connected/cli.py`).

Please uninstall not only `Pillow` but also `bimmer-connected`, from both your normal account and with `sudo`.

**Comment by Isy99** [comment]
Hi Richard,
got it running now.
I uninstalled and deleted pillow, pip and bimmer_connected.
Installed pip, that's to be found under pi's home .local

Installed dimmer-connected.
Perhaps that's the reason, why bimmer_connected then is to be found in pi's home path ?
Copied alle files under .local/lib and .local/bin into the corresponding directories under /usr

Well, not the best way, I guess. But it works,  until I install a brand new PI system with bookworm.

Kind regards,
Helmut
